// @hellomoon/api/dist/api/requests/Sacrifice{dropId}Request.class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sacrifice = void 0;
const BaseRequest_class_1 = require("../BaseRequest.class");
{
}
RequestArgs;
{
    /**
     * The pubkey of the user making the request
     */
    pubKey ?  : string;
    requestSigned ?  : {
        signature: {
            /**
         * The type of signature, usually buffer
         */
            type: string,
            /**
         * The signature data of the user signing the data
         */
            data: number[]
        } |
            string,
        message: {
            pubKey: string,
            oogyMint: string,
            oozeMint: string
        }
    };
    transactionSigned ?  : {
        type: string,
        data: number[]
    };
}
class Sacrifice {
    dropId;
}
exports.Sacrifice = Sacrifice;
Request;
BaseRequest_class_1.BaseRequest < Sacrifice;
{
    dropId;
}
RequestArgs, Response_interface_1.PaginatedResponse < Sacrifice;
{
    dropId;
}
 >> {
    method = "POST",
    url = "/v0/hello-moon/oogy/sacrifice/{dropId}",
    constructor(body, { dropId }, RequestArgs = {}) {
        super(body);
    }
};
