// @hellomoon/api/dist/api/requests/SpiceSacrifice{dropId}Request.class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpiceSacrifice = void 0;
const BaseRequest_class_1 = require("../BaseRequest.class");
{
}
RequestArgs;
{
    requestSigned ?  : {
        signature: {
            /**
         * The type of signature, usually buffer
         */
            type: string,
            /**
         * The signature data of the user signing the data
         */
            data: number[]
        } |
            string,
        message: {
            pubKey: string,
            oogyMint: string,
            numSpice: number,
            timeStamp: number
        }
    };
}
class SpiceSacrifice {
    dropId;
}
exports.SpiceSacrifice = SpiceSacrifice;
Request;
BaseRequest_class_1.BaseRequest < SpiceSacrifice;
{
    dropId;
}
RequestArgs, Response_interface_1.PaginatedResponse < SpiceSacrifice;
{
    dropId;
}
 >> {
    method = "POST",
    url = "/v0/hello-moon/oogy/spice-sacrifice/{dropId}",
    constructor(body, { dropId }, RequestArgs = {}) {
        super(body);
    }
};
