// @shyft-to/js/dist/esm/types/Callback.js
export var TxnAction;
(function (TxnAction) {
    TxnAction["NFT_MINT"] = "NFT_MINT";
    TxnAction["TOKEN_MINT"] = "TOKEN_MINT";
    TxnAction["TOKEN_CREATE"] = "TOKEN_CREATE";
    TxnAction["SOL_TRANSFER"] = "SOL_TRANSFER";
    TxnAction["TOKEN_TRANSFER"] = "TOKEN_TRANSFER";
    TxnAction["NFT_TRANSFER"] = "NFT_TRANSFER";
    TxnAction["NFT_BURN"] = "NFT_BURN";
    TxnAction["TOKEN_BURN"] = "TOKEN_BURN";
    TxnAction["NFT_SALE"] = "NFT_SALE";
    TxnAction["NFT_BID"] = "NFT_BID";
    TxnAction["NFT_BID_CANCEL"] = "NFT_BID_CANCEL";
    TxnAction["NFT_LIST"] = "NFT_LIST";
    TxnAction["NFT_LIST_UPDATE"] = "NFT_LIST_UPDATE";
    TxnAction["NFT_LIST_CANCEL"] = "NFT_LIST_CANCEL";
    TxnAction["MARKETPLACE_WITHDRAW"] = "MARKETPLACE_WITHDRAW";
    TxnAction["OFFER_LOAN"] = "OFFER_LOAN";
    TxnAction["CANCEL_LOAN"] = "CANCEL_LOAN";
    TxnAction["TAKE_LOAN"] = "TAKE_LOAN";
    TxnAction["REPAY_LOAN"] = "REPAY_LOAN";
    TxnAction["FORECLOSE_LOAN"] = "FORECLOSE_LOAN";
    TxnAction["REPAY_ESCROW_LOAN"] = "REPAY_ESCROW_LOAN";
    TxnAction["EXTEND_LOAN"] = "EXTEND_LOAN";
    TxnAction["EXTEND_ESCROW_LOAN"] = "EXTEND_ESCROW_LOAN";
    TxnAction["MEMO"] = "MEMO";
    TxnAction["SWAP"] = "SWAP";
    TxnAction["CREATE_POOL"] = "CREATE_POOL";
    TxnAction["ADD_LIQUIDITY"] = "ADD_LIQUIDITY";
    TxnAction["REMOVE_LIQUIDITY"] = "REMOVE_LIQUIDITY";
    TxnAction["CREATE_RAFFLE"] = "CREATE_RAFFLE";
    TxnAction["BUY_TICKETS"] = "BUY_TICKETS";
    TxnAction["REVEAL_WINNERS"] = "REVEAL_WINNERS";
    TxnAction["CLAIM_PRIZE"] = "CLAIM_PRIZE";
    TxnAction["CLOSE_RAFFLE"] = "CLOSE_RAFFLE";
    TxnAction["CANCEL_RAFFLE"] = "CANCEL_RAFFLE";
    TxnAction["CREATE_TREE"] = "CREATE_TREE";
    TxnAction["COMPRESSED_NFT_MINT"] = "COMPRESSED_NFT_MINT";
    TxnAction["COMPRESSED_NFT_TRANSFER"] = "COMPRESSED_NFT_TRANSFER";
    TxnAction["COMPRESSED_NFT_BURN"] = "COMPRESSED_NFT_BURN";
    TxnAction["CREATE_REALM"] = "CREATE_REALM";
    TxnAction["DEPOSIT_GOVERNING_TOKENS"] = "DEPOSIT_GOVERNING_TOKENS";
    TxnAction["WITHDRAW_GOVERNING_TOKENS"] = "WITHDRAW_GOVERNING_TOKENS";
    TxnAction["SET_GOVERNANCE_DELEGATE"] = "SET_GOVERNANCE_DELEGATE";
    TxnAction["CREATE_GOVERNANCE"] = "CREATE_GOVERNANCE";
    TxnAction["CREATE_PROGRAM_GOVERNANCE"] = "CREATE_PROGRAM_GOVERNANCE";
    TxnAction["CREATE_PROPOSAL"] = "CREATE_PROPOSAL";
    TxnAction["ADD_SIGNATORY"] = "ADD_SIGNATORY";
    TxnAction["REMOVE_SIGNATORY"] = "REMOVE_SIGNATORY";
    TxnAction["INSERT_TRANSACTION"] = "INSERT_TRANSACTION";
    TxnAction["REMOVE_TRANSACTION"] = "REMOVE_TRANSACTION";
    TxnAction["CANCEL_PROPOSAL"] = "CANCEL_PROPOSAL";
    TxnAction["SIGN_OFF_PROPOSAL"] = "SIGN_OFF_PROPOSAL";
    TxnAction["CAST_VOTE"] = "CAST_VOTE";
    TxnAction["FINALIZE_VOTE"] = "FINALIZE_VOTE";
    TxnAction["RELINQUISH_VOTE"] = "RELINQUISH_VOTE";
    TxnAction["EXECUTE_TRANSACTION"] = "EXECUTE_TRANSACTION";
    TxnAction["CREATE_MINT_GOVERNANCE"] = "CREATE_MINT_GOVERNANCE";
    TxnAction["CREATE_TOKEN_GOVERNANCE"] = "CREATE_TOKEN_GOVERNANCE";
    TxnAction["SET_GOVERNANCE_CONFIG"] = "SET_GOVERNANCE_CONFIG";
    TxnAction["SET_REALM_AUTHORITY"] = "SET_REALM_AUTHORITY";
    TxnAction["SET_REALM_CONFIG"] = "SET_REALM_CONFIG";
    TxnAction["CREATE_TOKEN_OWNER_RECORD"] = "CREATE_TOKEN_OWNER_RECORD";
    TxnAction["POST_MESSAGE"] = "POST_MESSAGE";
    TxnAction["UNKNOWN"] = "UNKNOWN";
})(TxnAction || (TxnAction = {}));
