// @shyft-to/js/dist/esm/types/DAS.js
export var Interface;
(function (Interface) {
    Interface["V1NFT"] = "V1_NFT";
    Interface["CUSTOM"] = "Custom";
    Interface["V1PRINT"] = "V1_PRINT";
    Interface["LEGACYNFT"] = "Legacy_NFT";
    Interface["V2NFT"] = "V2_NFT";
    Interface["FUNGIBLE_ASSET"] = "FungibleAsset";
    Interface["IDENTITY"] = "Identity";
    Interface["EXECUTABLE"] = "Executable";
    Interface["PROGRAMMABLENFT"] = "ProgrammableNFT";
})(Interface || (Interface = {}));
export var Scope;
(function (Scope) {
    Scope["FULL"] = "full";
    Scope["ROYALTY"] = "royalty";
    Scope["METADATA"] = "metadata";
    Scope["EXTENSION"] = "extension";
})(Scope || (Scope = {}));
export var Context;
(function (Context) {
    Context["WalletDefault"] = "wallet-default";
    Context["WebDesktop"] = "web-desktop";
    Context["WebMobile"] = "web-mobile";
    Context["AppMobile"] = "app-mobile";
    Context["AppDesktop"] = "app-desktop";
    Context["App"] = "app";
    Context["Vr"] = "vr";
})(Context || (Context = {}));
export var OwnershipModel;
(function (OwnershipModel) {
    OwnershipModel["SINGLE"] = "single";
    OwnershipModel["TOKEN"] = "token";
})(OwnershipModel || (OwnershipModel = {}));
export var RoyaltyModel;
(function (RoyaltyModel) {
    RoyaltyModel["CREATORS"] = "creators";
    RoyaltyModel["FANOUT"] = "fanout";
    RoyaltyModel["SINGLE"] = "single";
})(RoyaltyModel || (RoyaltyModel = {}));
export var UseMethods;
(function (UseMethods) {
    UseMethods["BURN"] = "Burn";
    UseMethods["SINGLE"] = "Single";
    UseMethods["MULTIPLE"] = "Multiple";
})(UseMethods || (UseMethods = {}));
