// @hellomoon/api/dist/api/BaseRequest.class.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseRequest = void 0;
const axios_1 = __importDefault(require("axios"));
class BaseRequest {
    body;
    params;
    headers;
    customParams;
    static baseUrl = "https://rest-api.hellomoon.io";
    constructor(body = undefined, params = {}, headers = {}, customParams = "") {
        this.body = body;
        this.params = params;
        this.headers = headers;
        this.customParams = customParams;
    }
    getRequestUrl(baseUrl) {
        return (baseUrl || BaseRequest.baseUrl) + this.url;
    }
    async sendAsCustomUrlParam(apiKey, baseUrl) {
        const res = await (0, axios_1.default)({
            method: this.method,
            url: this.getRequestUrl(baseUrl) + "/" + this.customParams,
            headers: {
                ...this.headers,
                Authorization: `Bearer ${apiKey}`,
            },
        });
        return res.data;
    }
    async _send(apiKey, baseUrl) {
        const res = await (0, axios_1.default)({
            method: this.method,
            data: this.body,
            url: this.params
                ? interpolateParams(this.getRequestUrl(baseUrl), this.params)
                : this.getRequestUrl(baseUrl),
            headers: {
                ...this.headers,
                Authorization: `Bearer ${apiKey}`,
            },
        });
        return res.data;
    }
}
exports.BaseRequest = BaseRequest;
function interpolateParams(url, params) {
    let interpolatedUrl = url;
    Object.entries(params).map(([k, v]) => {
        interpolatedUrl = interpolatedUrl.replace(`{${k}}`, v.toString());
    });
    return interpolatedUrl;
}
