// @shyft-to/js/dist/cjs/types/DAS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UseMethods = exports.RoyaltyModel = exports.OwnershipModel = exports.Context = exports.Scope = exports.Interface = void 0;
var Interface;
(function (Interface) {
    Interface["V1NFT"] = "V1_NFT";
    Interface["CUSTOM"] = "Custom";
    Interface["V1PRINT"] = "V1_PRINT";
    Interface["LEGACYNFT"] = "Legacy_NFT";
    Interface["V2NFT"] = "V2_NFT";
    Interface["FUNGIBLE_ASSET"] = "FungibleAsset";
    Interface["IDENTITY"] = "Identity";
    Interface["EXECUTABLE"] = "Executable";
    Interface["PROGRAMMABLENFT"] = "ProgrammableNFT";
})(Interface = exports.Interface || (exports.Interface = {}));
var Scope;
(function (Scope) {
    Scope["FULL"] = "full";
    Scope["ROYALTY"] = "royalty";
    Scope["METADATA"] = "metadata";
    Scope["EXTENSION"] = "extension";
})(Scope = exports.Scope || (exports.Scope = {}));
var Context;
(function (Context) {
    Context["WalletDefault"] = "wallet-default";
    Context["WebDesktop"] = "web-desktop";
    Context["WebMobile"] = "web-mobile";
    Context["AppMobile"] = "app-mobile";
    Context["AppDesktop"] = "app-desktop";
    Context["App"] = "app";
    Context["Vr"] = "vr";
})(Context = exports.Context || (exports.Context = {}));
var OwnershipModel;
(function (OwnershipModel) {
    OwnershipModel["SINGLE"] = "single";
    OwnershipModel["TOKEN"] = "token";
})(OwnershipModel = exports.OwnershipModel || (exports.OwnershipModel = {}));
var RoyaltyModel;
(function (RoyaltyModel) {
    RoyaltyModel["CREATORS"] = "creators";
    RoyaltyModel["FANOUT"] = "fanout";
    RoyaltyModel["SINGLE"] = "single";
})(RoyaltyModel = exports.RoyaltyModel || (exports.RoyaltyModel = {}));
var UseMethods;
(function (UseMethods) {
    UseMethods["BURN"] = "Burn";
    UseMethods["SINGLE"] = "Single";
    UseMethods["MULTIPLE"] = "Multiple";
})(UseMethods = exports.UseMethods || (exports.UseMethods = {}));
