// @shyft-to/js/dist/cjs/api/mp-bidding-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MpBiddingClient = void 0;
const utils_1 = require("../utils");
class MpBiddingClient {
    constructor(config) {
        this.config = config;
    }
    active(input) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const params = {
                    network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                    marketplace_address: input.marketplaceAddress,
                };
                if (input === null || input === void 0 ? void 0 : input.buyerAddress) {
                    params['buyer_address'] = input.buyerAddress;
                }
                if (input === null || input === void 0 ? void 0 : input.collectionAddress) {
                    params['collection_address'] = input.collectionAddress;
                }
                if (input === null || input === void 0 ? void 0 : input.nftAddress) {
                    params['nft_address'] = input.nftAddress;
                }
                if (input === null || input === void 0 ? void 0 : input.sortBy) {
                    params['sort_by'] = input.sortBy;
                }
                if (input === null || input === void 0 ? void 0 : input.sortOrder) {
                    params['sort_order'] = input.sortOrder;
                }
                if (input === null || input === void 0 ? void 0 : input.page) {
                    params['page'] = input.page;
                }
                if (input === null || input === void 0 ? void 0 : input.size) {
                    params['size'] = input.size;
                }
                const data = yield (0, utils_1.restApiCall)(this.config.apiKey, {
                    method: 'get',
                    url: 'marketplace/active_bids',
                    params,
                });
                const bids = data.result;
                bids.data.forEach((x) => {
                    x.created_at = new Date(x.created_at);
                });
                return bids;
            }
            catch (error) {
                throw error;
            }
        });
    }
    bid(input) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const reqBody = {
                    network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                    marketplace_address: input.marketplaceAddress,
                    nft_address: input.nftAddress,
                    price: input.price,
                    buyer_wallet: input.buyerWallet,
                };
                if (input === null || input === void 0 ? void 0 : input.isGasLess) {
                    reqBody['on_the_house'] = input.isGasLess;
                }
                if (input === null || input === void 0 ? void 0 : input.serviceCharge) {
                    reqBody['service_charge'] = input.serviceCharge;
                }
                const data = yield (0, utils_1.restApiCall)(this.config.apiKey, {
                    method: 'post',
                    url: 'marketplace/bid',
                    data: reqBody,
                });
                const response = data.result;
                return response;
            }
            catch (error) {
                throw error;
            }
        });
    }
    cancelBid(input) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const reqBody = {
                    network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                    marketplace_address: input.marketplaceAddress,
                    bid_state: input.bidState,
                    buyer_wallet: input.buyerWallet,
                };
                if (input === null || input === void 0 ? void 0 : input.isGasLess) {
                    reqBody['on_the_house'] = input.isGasLess;
                }
                const data = yield (0, utils_1.restApiCall)(this.config.apiKey, {
                    method: 'post',
                    url: 'marketplace/cancel_bid',
                    data: reqBody,
                });
                const encodedTransaction = data.result.encoded_transaction;
                return encodedTransaction;
            }
            catch (error) {
                throw error;
            }
        });
    }
    acceptBid(input) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const reqBody = {
                    network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                    marketplace_address: input.marketplaceAddress,
                    bid_state: input.bidState,
                    seller_wallet: input.sellerWallet,
                };
                const data = yield (0, utils_1.restApiCall)(this.config.apiKey, {
                    method: 'post',
                    url: 'marketplace/accept_bid',
                    data: reqBody,
                });
                const response = data.result;
                return response;
            }
            catch (error) {
                throw error;
            }
        });
    }
}
exports.MpBiddingClient = MpBiddingClient;
