// @shyft-to/js/dist/cjs/utils/shyft.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShyftSdk = void 0;
const api_1 = require("../api");
const utils_1 = require("../utils");
const semi_custodial_wallet_client_1 = require("../api/semi-custodial-wallet-client");
const web3_js_1 = require("@solana/web3.js");
const utility_1 = require("./utility");
class ShyftSdk {
    constructor(settings) {
        this.config = new utils_1.ShyftConfig(settings);
        this.connection = new web3_js_1.Connection((0, utility_1.shyftClusterApiUrl)(this.config.apiKey, this.config.network));
        this.wallet = new api_1.WalletClient(this.config);
        this.nft = new api_1.NftClient(this.config);
        this.token = new api_1.TokenClient(this.config);
        this.candyMachine = new api_1.CandyMachineClient(this.config);
        this.marketplace = new api_1.MarketplaceClient(this.config);
        this.transaction = new api_1.TransactionClient(this.config);
        this.txnRelayer = new api_1.TxnRelayerClient(this.config);
        this.storage = new api_1.StorageClient(this.config);
        this.semiCustodialWallet = new semi_custodial_wallet_client_1.SemiCustodialWalletClient(this.config);
        this.callback = new api_1.CallbackClient(this.config);
        this.rpc = new api_1.RpcClient(this.connection);
    }
}
exports.ShyftSdk = ShyftSdk;
