// @hellomoon/api/dist/subscriptions/streams/buddy-link-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuddyLinkStream = void 0;
const base_data_stream_class_1 = require("../models/base-data-stream.class");
const fieldMap = {
    blockId: { type: "NUMERIC" },
    blockTime: { type: "DATE" },
    transactionId: { type: "TEXT" },
    transactionPosition: { type: "NUMERIC" },
    instructionOrdinal: { type: "NUMERIC" },
    instructionName: { type: "TEXT" },
    eventType: { type: "ENUM" },
    "buddyLinkEvent.buddyProgram": { type: "TEXT" },
    "buddyLinkEvent.superAdmin": { type: "TEXT" },
    "buddyLinkEvent.authority": { type: "TEXT" },
    "buddyLinkEvent.buddy": { type: "TEXT" },
    "buddyLinkEvent.referrerSharedTreasury": { type: "TEXT" },
    "buddyLinkEvent.buddyName": { type: "TEXT" },
    "buddyLinkEvent.payer": { type: "TEXT" },
    "buddyLinkEvent.buddyProfile": { type: "TEXT" },
    "buddyLinkEvent.usdcMint": { type: "TEXT" },
    "buddyLinkEvent.usdcTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.buddylinkUsdcTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.referrerUsdcSharedTreasury": { type: "TEXT" },
    "buddyLinkEvent.referrerUsdcTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.owner1": { type: "TEXT" },
    "buddyLinkEvent.owner2": { type: "TEXT" },
    "buddyLinkEvent.owner3": { type: "TEXT" },
    "buddyLinkEvent.owner4": { type: "TEXT" },
    "buddyLinkEvent.mint": { type: "TEXT" },
    "buddyLinkEvent.buddyTreasury": { type: "TEXT" },
    "buddyLinkEvent.buddyTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.organization": { type: "TEXT" },
    "buddyLinkEvent.globalReferrerTreasury": { type: "TEXT" },
    "buddyLinkEvent.ownerTreasury": { type: "TEXT" },
    "buddyLinkEvent.referrerTreasury": { type: "TEXT" },
    "buddyLinkEvent.member": { type: "TEXT" },
    "buddyLinkEvent.ownerTreasuryTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.referrerTreasuryTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.referrerBuddy1": { type: "TEXT" },
    "buddyLinkEvent.referrerBuddy2": { type: "TEXT" },
    "buddyLinkEvent.referrerBuddy3": { type: "TEXT" },
    "buddyLinkEvent.referrerBuddy4": { type: "TEXT" },
    "buddyLinkEvent.memberStatistics": { type: "TEXT" },
    "buddyLinkEvent.ambassador": { type: "TEXT" },
    "buddyLinkEvent.masterOrganization": { type: "TEXT" },
    "buddyLinkEvent.blNftSigner": { type: "TEXT" },
    "buddyLinkEvent.masterOrganizationTreasury": { type: "TEXT" },
    "buddyLinkEvent.masterOrganizationTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.organizationName": { type: "TEXT" },
    "buddyLinkEvent.buddylinkBuddy": { type: "TEXT" },
    "buddyLinkEvent.buddylinkBuddyTreasury": { type: "TEXT" },
    "buddyLinkEvent.buddylinkBuddyTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.buddylinkBuddyMember": { type: "TEXT" },
    "buddyLinkEvent.userTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.receiverAccount": { type: "TEXT" },
    "buddyLinkEvent.referrerTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.ambassadorClaimed": { type: "TEXT" },
    "buddyLinkEvent.ambassadorTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.buddyLinkTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.refereeTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.referrerMember": { type: "TEXT" },
    "buddyLinkEvent.referrerTreasuryForReward": { type: "TEXT" },
    "buddyLinkEvent.buddyGlobalReferrerTreasury": { type: "TEXT" },
    "buddyLinkEvent.buddyGlobalReferrerTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.globalReferrerTreasuryForReward": { type: "TEXT" },
    "buddyLinkEvent.globalReferrerTokenAccount": { type: "TEXT" },
    "buddyLinkEvent.programSigner": { type: "TEXT" },
    "buddyLinkEvent.funder": { type: "TEXT" },
    "buddyLinkEvent.referrerMemberStatistics": { type: "TEXT" },
    "buddyLinkEvent.admin": { type: "TEXT" },
    "buddyLinkEvent.newAuthority": { type: "TEXT" },
    "buddyLinkEvent.firstAuthorizedAuthority": { type: "TEXT" },
    "buddyLinkEvent.secondAuthorizedAuthority": { type: "TEXT" },
    "buddyLinkEvent.toSetVersion": { type: "TEXT" },
    "buddyLinkEvent.name": { type: "TEXT" },
    "buddyLinkEvent.ambassadorShareKeptInBps": { type: "NUMERIC" },
    "buddyLinkEvent.mainTokenMint": { type: "TEXT" },
    "buddyLinkEvent.enforceWalletUniqueness": { type: "BOOLEAN" },
    "buddyLinkEvent.allowMultiLevel": { type: "BOOLEAN" },
    "buddyLinkEvent.ambassadorBps": { type: "NUMERIC" },
    "buddyLinkEvent.allowMultiAttribution": { type: "BOOLEAN" },
    "buddyLinkEvent.volumeRequiredForRewards": { type: "NUMERIC" },
    "buddyLinkEvent.amount": { type: "NUMERIC" },
    "buddyLinkEvent.ambassadorClaimAmount": { type: "NUMERIC" },
    "buddyLinkEvent.transferRewardNoMultiLevelAmount": { type: "NUMERIC" },
    "buddyLinkEvent.transferRewardCheckedAmount": { type: "NUMERIC" },
    "buddyLinkEvent.transferRewardGlobalAmount": { type: "NUMERIC" },
    "buddyLinkEvent.totalAmount": { type: "NUMERIC" },
    "buddyLinkEvent.transferRewardSplAmount": { type: "NUMERIC" },
    "buddyLinkEvent.membersIncluded": { type: "BOOLEAN" },
    "buddyLinkEvent.transferRewardSecureNoGlobalAmount": { type: "NUMERIC" },
    "buddyLinkEvent.transferRewardStarAtlasAmount": { type: "NUMERIC" },
    "buddyLinkEvent.volume": { type: "NUMERIC" },
    "buddyLinkEvent.shareKeptInBps": { type: "NUMERIC" },
    "buddyLinkEvent.configurationValue": { type: "TEXT" },
    "buddyLinkEvent.isEnabled": { type: "BOOLEAN" },
    "buddyLinkEvent.volumeRequiredForReward": { type: "NUMERIC" },
    "buddyLinkEvent.transferRewardCheckedTestingAmount": { type: "NUMERIC" },
};
class BuddyLinkStream extends base_data_stream_class_1.BaseDataStream {
    id = "BuddyLink";
    static fieldMap = fieldMap;
}
exports.BuddyLinkStream = BuddyLinkStream;
