// @hellomoon/api/dist/subscriptions/streams/lp-deposit-withdrawal-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LpDepositWithdrawalStream = void 0;
const base_data_stream_class_1 = require("../models/base-data-stream.class");
const fieldMap = {
    blockId: { type: "NUMERIC" },
    blockTime: { type: "DATE" },
    transactionId: { type: "TEXT" },
    transactionPosition: { type: "NUMERIC" },
    instructionOrdinal: { type: "NUMERIC" },
    programId: { type: "TEXT" },
    instructionName: { type: "TEXT" },
    poolAddress: { type: "TEXT" },
    lpTokenMint: { type: "TEXT" },
    userAccount: { type: "TEXT" },
    actionType: { type: "ENUM" },
    tokenMint: { type: "COMPOSITE", dataType: "TEXT", fields: ["mintTokenA", "mintTokenB"] },
    tokenAmount: { type: "COMPOSITE", dataType: "TEXT", fields: ["amountTokenA", "amountTokenB"] },
};
class LpDepositWithdrawalStream extends base_data_stream_class_1.BaseDataStream {
    id = "LpDepositWithdrawal";
    static fieldMap = fieldMap;
}
exports.LpDepositWithdrawalStream = LpDepositWithdrawalStream;
