// @hellomoon/api/dist/subscriptions/streams/nft-primary-sale-with-collection-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftPrimarySaleWithCollectionStream = void 0;
const base_data_stream_class_1 = require("../models/base-data-stream.class");
const fieldMap = {
    helloMoonCollectionId: { type: "TEXT" },
    collectionName: { type: "TEXT" },
    programId: { type: "TEXT" },
    instructionName: { type: "TEXT" },
    price: { type: "NUMERIC" },
    payerAccount: { type: "TEXT" },
    isSmartMinter: { type: "BOOLEAN" },
    sellerAccount: { type: "TEXT" },
    mint: { type: "TEXT" },
    blockTime: { type: "DATE" },
    blockId: { type: "NUMERIC" },
    transactionId: { type: "TEXT" },
    transactionPosition: { type: "NUMERIC" },
    instructionOrdinal: { type: "NUMERIC" },
};
class NftPrimarySaleWithCollectionStream extends base_data_stream_class_1.BaseDataStream {
    id = "NftPrimarySaleWithCollection";
    static fieldMap = fieldMap;
}
exports.NftPrimarySaleWithCollectionStream = NftPrimarySaleWithCollectionStream;
