// @hellomoon/api/dist/subscriptions/streams/sharky-event-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharkyEventStream = void 0;
const base_data_stream_class_1 = require("../models/base-data-stream.class");
const fieldMap = {
    blockId: { type: "NUMERIC" },
    blockTime: { type: "DATE" },
    transactionId: { type: "TEXT" },
    transactionPosition: { type: "NUMERIC" },
    instructionOrdinal: { type: "NUMERIC" },
    eventType: { type: "ENUM" },
    floorPrice: { type: "NUMERIC" },
    "nftList.nftList": { type: "TEXT" },
    "nftList.payer": { type: "TEXT" },
    "nftList.collectionName": { type: "TEXT" },
    "orderBook.orderBook": { type: "TEXT" },
    "orderBook.payer": { type: "TEXT" },
    "orderBook.nftList": { type: "TEXT" },
    "orderBook.apy": { type: "NUMERIC" },
    "orderBook.loanDuration": { type: "NUMERIC" },
    "orderBook.feePermillicentage": { type: "NUMERIC" },
    "orderBook.orderBookFeeAuthority": { type: "TEXT" },
    "loan.lender": { type: "TEXT" },
    "loan.lenderValueTokenAccount": { type: "TEXT" },
    "loan.borrower": { type: "TEXT" },
    "loan.borrowerValueTokenAccount": { type: "TEXT" },
    "loan.borrowerCollateralAccount": { type: "TEXT" },
    "loan.valueMint": { type: "TEXT" },
    "loan.collateralMint": { type: "TEXT" },
    "loan.loan": { type: "TEXT" },
    "loan.escrow": { type: "TEXT" },
    "loan.escrowTokenAccount": { type: "TEXT" },
    "loan.escrowCollateralTokenAccount": { type: "TEXT" },
    "loan.orderBook": { type: "TEXT" },
    "loan.nftList": { type: "TEXT" },
    "loan.principalLamports": { type: "NUMERIC" },
    "loan.loanDuration": { type: "NUMERIC" },
    "loan.amount": { type: "NUMERIC" },
    "loan.nftListIndex": { type: "NUMERIC" },
    "loan.lenderWallet": { type: "TEXT" },
    "loan.offerTime": { type: "NUMERIC" },
    "loan.valueTokenMint": { type: "TEXT" },
    "loan.feeTokenAccount": { type: "TEXT" },
    "loan.feeAuthority": { type: "TEXT" },
    "loan.feeAmount": { type: "NUMERIC" },
    "loan.fundee": { type: "TEXT" },
    "loan.lenderCollateralAccount": { type: "TEXT" },
    "loan.collector": { type: "TEXT" },
};
class SharkyEventStream extends base_data_stream_class_1.BaseDataStream {
    id = "SharkyEvent";
    static fieldMap = fieldMap;
}
exports.SharkyEventStream = SharkyEventStream;
