// @hellomoon/api/dist/subscriptions/streams/star-atlas-event-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StarAtlasEventStream = void 0;
const base_data_stream_class_1 = require("../models/base-data-stream.class");
const fieldMap = {
    blockId: { type: "NUMERIC" },
    blockTime: { type: "DATE" },
    transactionId: { type: "TEXT" },
    transactionPosition: { type: "NUMERIC" },
    instructionOrdinal: { type: "NUMERIC" },
    instructionName: { type: "TEXT" },
    eventType: { type: "ENUM" },
    "escapeVelocityEvent.owner": { type: "TEXT" },
    "escapeVelocityEvent.game": { type: "TEXT" },
    "escapeVelocityEvent.fleetOwner": { type: "TEXT" },
    "escapeVelocityEvent.fleet": { type: "TEXT" },
    "escapeVelocityEvent.funder": { type: "TEXT" },
    "escapeVelocityEvent.userTokenAccount": { type: "TEXT" },
    "escapeVelocityEvent.fleetSigner": { type: "TEXT" },
    "escapeVelocityEvent.fleetLockAccount": { type: "TEXT" },
    "escapeVelocityEvent.lockMint": { type: "TEXT" },
    "escapeVelocityEvent.lockAmount": { type: "TEXT" },
    "escapeVelocityEvent.representation": { type: "TEXT" },
    "escapeVelocityEvent.defaultStats.warpSpeed": { type: "NUMERIC" },
    "escapeVelocityEvent.defaultStats.warpCooldown": { type: "NUMERIC" },
    "escapeVelocityEvent.defaultStats.warpRange": { type: "NUMERIC" },
    "escapeVelocityEvent.to": { type: "NUMERIC" },
    "escapeVelocityEvent.fleetAmount": { type: "NUMERIC" },
    "galacticMarketplaceEvent.updateAuthorityMaster": { type: "TEXT" },
    "galacticMarketplaceEvent.funder": { type: "TEXT" },
    "galacticMarketplaceEvent.marketVarsAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.feeExemptTarget": { type: "TEXT" },
    "galacticMarketplaceEvent.feeExemptAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.updateAuthorityAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.registeredCurrency": { type: "TEXT" },
    "galacticMarketplaceEvent.currencyMint": { type: "TEXT" },
    "galacticMarketplaceEvent.payer": { type: "TEXT" },
    "galacticMarketplaceEvent.user": { type: "TEXT" },
    "galacticMarketplaceEvent.openOrdersCounter": { type: "TEXT" },
    "galacticMarketplaceEvent.depositMint": { type: "TEXT" },
    "galacticMarketplaceEvent.signer": { type: "TEXT" },
    "galacticMarketplaceEvent.orderInitializer": { type: "TEXT" },
    "galacticMarketplaceEvent.initializerDepositTokenAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.orderVaultAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.orderVaultAuthority": { type: "TEXT" },
    "galacticMarketplaceEvent.orderAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.orderTaker": { type: "TEXT" },
    "galacticMarketplaceEvent.orderTakerDepositTokenAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.orderTakerReceiveTokenAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.assetMint": { type: "TEXT" },
    "galacticMarketplaceEvent.initializerReceiveTokenAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.saVault": { type: "TEXT" },
    "galacticMarketplaceEvent.atlasStaking": { type: "TEXT" },
    "galacticMarketplaceEvent.registeredStake": { type: "TEXT" },
    "galacticMarketplaceEvent.stakingAccount": { type: "TEXT" },
    "galacticMarketplaceEvent.receiveMint": { type: "TEXT" },
    "galacticMarketplaceEvent.saCurrencyVault": { type: "TEXT" },
    "galacticMarketplaceEvent.discount": { type: "NUMERIC" },
    "galacticMarketplaceEvent.stakeAmount": { type: "NUMERIC" },
    "galacticMarketplaceEvent.purchaseQuantity": { type: "NUMERIC" },
    "galacticMarketplaceEvent.expectedPrice": { type: "NUMERIC" },
    "galacticMarketplaceEvent.seller": { type: "TEXT" },
    "galacticMarketplaceEvent.price": { type: "NUMERIC" },
    "galacticMarketplaceEvent.originationQty": { type: "NUMERIC" },
    "galacticMarketplaceEvent.royalty": { type: "NUMERIC" },
    "galacticMarketplaceEvent.amount": { type: "NUMERIC" },
    "galacticMarketplaceEvent.sellAmount": { type: "NUMERIC" },
    "galacticMarketplaceEvent.sellMint": { type: "TEXT" },
    "galacticMarketplaceEvent.buyAmount": { type: "NUMERIC" },
    "galacticMarketplaceEvent.buyMint": { type: "TEXT" },
    "claimStakesEvent.playerAccount": { type: "TEXT" },
    "claimStakesEvent.claimStakingAccount": { type: "TEXT" },
    "claimStakesEvent.claimStakeVarsAccount": { type: "TEXT" },
    "claimStakesEvent.fuelTreasuryTokenAccount": { type: "TEXT" },
    "claimStakesEvent.armsTreasuryTokenAccount": { type: "TEXT" },
    "claimStakesEvent.foodTreasuryTokenAccount": { type: "TEXT" },
    "claimStakesEvent.toolkitTreasuryTokenAccount": { type: "TEXT" },
    "claimStakesEvent.playerFuelTokenAccount": { type: "TEXT" },
    "claimStakesEvent.playerArmsTokenAccount": { type: "TEXT" },
    "claimStakesEvent.playerFoodTokenAccount": { type: "TEXT" },
    "claimStakesEvent.playerToolkitTokenAccount": { type: "TEXT" },
    "claimStakesEvent.treasuryAuthorityAccount": { type: "TEXT" },
    "claimStakesEvent.claimStakeMint": { type: "TEXT" },
    "claimStakesEvent.funder": { type: "TEXT" },
    "claimStakesEvent.claimStakeVars": { type: "TEXT" },
    "claimStakesEvent.playerFactionAccount": { type: "TEXT" },
    "claimStakesEvent.escrowAuthority": { type: "TEXT" },
    "claimStakesEvent.claimStakeTokenAccountSource": { type: "TEXT" },
    "claimStakesEvent.claimStakeTokenAccountEscrow": { type: "TEXT" },
    "claimStakesEvent.updateAuthorityAccount": { type: "TEXT" },
    "claimStakesEvent.globalVarsAccount": { type: "TEXT" },
    "claimStakesEvent.fuelTreasuryAccount": { type: "TEXT" },
    "claimStakesEvent.armsTreasuryAccount": { type: "TEXT" },
    "claimStakesEvent.foodTreasuryAccount": { type: "TEXT" },
    "claimStakesEvent.toolkitTreasuryAccount": { type: "TEXT" },
    "claimStakesEvent.fuelMint": { type: "TEXT" },
    "claimStakesEvent.foodMint": { type: "TEXT" },
    "claimStakesEvent.armsMint": { type: "TEXT" },
    "claimStakesEvent.toolkitMint": { type: "TEXT" },
    "claimStakesEvent.fromTokenAccount": { type: "TEXT" },
    "claimStakesEvent.claimStakeEscrow": { type: "TEXT" },
    "claimStakesEvent.playerTokenAccount": { type: "TEXT" },
    "claimStakesEvent.claimStakeEscrowAccount": { type: "TEXT" },
    "claimStakesEvent.claimStakeQuantity": { type: "NUMERIC" },
    "claimStakesEvent.rewardRatePerSecond.fuel": { type: "NUMERIC" },
    "claimStakesEvent.rewardRatePerSecond.arms": { type: "NUMERIC" },
    "claimStakesEvent.rewardRatePerSecond.food": { type: "NUMERIC" },
    "claimStakesEvent.rewardRatePerSecond.toolkit": { type: "NUMERIC" },
    "claimStakesEvent.claimStakeMaxReserves.fuel": { type: "NUMERIC" },
    "claimStakesEvent.claimStakeMaxReserves.arms": { type: "NUMERIC" },
    "claimStakesEvent.claimStakeMaxReserves.food": { type: "NUMERIC" },
    "claimStakesEvent.claimStakeMaxReserves.toolkit": { type: "NUMERIC" },
    "claimStakesEvent.rewardRateValues.fuel": { type: "NUMERIC" },
    "claimStakesEvent.rewardRateValues.arms": { type: "NUMERIC" },
    "claimStakesEvent.rewardRateValues.food": { type: "NUMERIC" },
    "claimStakesEvent.rewardRateValues.toolkit": { type: "NUMERIC" },
    "claimStakesEvent.maxReserveValues.fuel": { type: "NUMERIC" },
    "claimStakesEvent.maxReserveValues.arms": { type: "NUMERIC" },
    "claimStakesEvent.maxReserveValues.food": { type: "NUMERIC" },
    "claimStakesEvent.maxReserveValues.toolkit": { type: "NUMERIC" },
    "claimStakesEvent.fuelAmount": { type: "NUMERIC" },
    "claimStakesEvent.armsAmount": { type: "NUMERIC" },
    "claimStakesEvent.foodAmount": { type: "NUMERIC" },
    "claimStakesEvent.toolkitAmount": { type: "NUMERIC" },
    "factionEnlistmentEvent.playerFactionAccount": { type: "TEXT" },
    "factionEnlistmentEvent.playerAccount": { type: "TEXT" },
    "factionEnlistmentEvent.factionId": { type: "TEXT" },
    "scoresEvent.playerAccount": { type: "TEXT" },
    "scoresEvent.shipStakingAccount": { type: "TEXT" },
    "scoresEvent.scoreVarsAccount": { type: "TEXT" },
    "scoresEvent.shipTokenAccountEscrow": { type: "TEXT" },
    "scoresEvent.fuelTokenAccountEscrow": { type: "TEXT" },
    "scoresEvent.foodTokenAccountEscrow": { type: "TEXT" },
    "scoresEvent.armsTokenAccountEscrow": { type: "TEXT" },
    "scoresEvent.escrowAuthority": { type: "TEXT" },
    "scoresEvent.shipMint": { type: "TEXT" },
    "scoresEvent.fuelMint": { type: "TEXT" },
    "scoresEvent.foodMint": { type: "TEXT" },
    "scoresEvent.armsMint": { type: "TEXT" },
    "scoresEvent.updateAuthorityAccount": { type: "TEXT" },
    "scoresEvent.scoreVarsShipAccount": { type: "TEXT" },
    "scoresEvent.playerAtlasTokenAccount": { type: "TEXT" },
    "scoresEvent.treasuryTokenAccount": { type: "TEXT" },
    "scoresEvent.treasuryAuthorityAccount": { type: "TEXT" },
    "scoresEvent.playerFactionAccount": { type: "TEXT" },
    "scoresEvent.shipTokenAccountSource": { type: "TEXT" },
    "scoresEvent.atlasMint": { type: "TEXT" },
    "scoresEvent.toolkitMint": { type: "TEXT" },
    "scoresEvent.tokenOwnerAccount": { type: "TEXT" },
    "scoresEvent.armsTokenAccountSource": { type: "TEXT" },
    "scoresEvent.foodTokenAccountSource": { type: "TEXT" },
    "scoresEvent.fuelTokenAccountSource": { type: "TEXT" },
    "scoresEvent.toolkitTokenAccountSource": { type: "TEXT" },
    "scoresEvent.armsTokenAccountReturn": { type: "TEXT" },
    "scoresEvent.foodTokenAccountReturn": { type: "TEXT" },
    "scoresEvent.fuelTokenAccountReturn": { type: "TEXT" },
    "scoresEvent.shipTokenAccountReturn": { type: "TEXT" },
    "scoresEvent.stakingBump": { type: "TEXT" },
    "scoresEvent.scorevarsBump": { type: "TEXT" },
    "scoresEvent.shipBump": { type: "TEXT" },
    "scoresEvent.fuelBump": { type: "TEXT" },
    "scoresEvent.foodBump": { type: "TEXT" },
    "scoresEvent.armsBump": { type: "TEXT" },
    "scoresEvent.escrowAuthBump": { type: "TEXT" },
    "scoresEvent.scorevarsShipBump": { type: "TEXT" },
    "scoresEvent.treasuryBump": { type: "TEXT" },
    "scoresEvent.treasuryAuthBump": { type: "TEXT" },
    "scoresEvent.escrowBump": { type: "TEXT" },
    "scoresEvent.shipQuantity": { type: "NUMERIC" },
    "scoresEvent.armsQuantity": { type: "NUMERIC" },
    "scoresEvent.foodQuantity": { type: "NUMERIC" },
    "scoresEvent.rewardRatePerSecond": { type: "NUMERIC" },
    "scoresEvent.fuelMaxReserve": { type: "NUMERIC" },
    "scoresEvent.foodMaxReserve": { type: "NUMERIC" },
    "scoresEvent.armsMaxReserve": { type: "NUMERIC" },
    "scoresEvent.toolkitMaxReserve": { type: "NUMERIC" },
    "scoresEvent.millisecondsToBurnOneFuel": { type: "NUMERIC" },
    "scoresEvent.millisecondsToBurnOneFood": { type: "NUMERIC" },
    "scoresEvent.millisecondsToBurnOneArms": { type: "NUMERIC" },
    "scoresEvent.millisecondsToBurnOneToolkit": { type: "NUMERIC" },
    "scoresEvent.newRewardRatePerSecond": { type: "NUMERIC" },
    "scoresEvent.atlasAmount": { type: "NUMERIC" },
    "scoresEvent.armsBurned": { type: "NUMERIC" },
    "scoresEvent.fuelBurned": { type: "NUMERIC" },
    "scoresEvent.foodBurned": { type: "NUMERIC" },
    "scoresEvent.toolkitBurned": { type: "NUMERIC" },
    "playerProfileEvent.newMember": { type: "TEXT" },
    "playerProfileEvent.roleAccount": { type: "TEXT" },
    "playerProfileEvent.roleMembershipAccount": { type: "TEXT" },
    "playerProfileEvent.funder": { type: "TEXT" },
    "playerProfileEvent.profile": { type: "TEXT" },
    "playerProfileEvent.key": { type: "TEXT" },
    "playerProfileEvent.newRoleAccount": { type: "TEXT" },
    "playerProfileEvent.member": { type: "TEXT" },
    "playerProfileEvent.roleNameAccount": { type: "TEXT" },
    "playerProfileEvent.name": { type: "TEXT" },
    "playerProfileEvent.authorizer": { type: "TEXT" },
    "playerProfileEvent.role": { type: "TEXT" },
    "playerProfileEvent.keyIndex": { type: "NUMERIC" },
    "playerProfileEvent.keyIndexInMembershipAccount": { type: "NUMERIC" },
    "playerProfileEvent.keyAddIndex": { type: "NUMERIC" },
    "playerProfileEvent.keyPermissionsIndex": { type: "NUMERIC" },
    "playerProfileEvent.newKeyThreshold": { type: "TEXT" },
    "playerProfileEvent.keyThreshold": { type: "TEXT" },
    "playerProfileEvent.keyIndexInRoleAccount": { type: "NUMERIC" },
    "playerProfileEvent.roleNameBump": { type: "TEXT" },
    "profileFactionEvent.key": { type: "TEXT" },
    "profileFactionEvent.funder": { type: "TEXT" },
    "profileFactionEvent.profile": { type: "TEXT" },
    "profileFactionEvent.faction": { type: "TEXT" },
    "profileFactionEvent.keyIndex": { type: "NUMERIC" },
    "profileFactionEvent.chooseFactionFaction": { type: "ENUM" },
    "atlasFeePayerEvent.funderPayment": { type: "NUMERIC" },
    "atlasFeePayerEvent.feePayer": { type: "TEXT" },
    "atlasFeePayerEvent.paymentAccount": { type: "TEXT" },
    "atlasFeePayerEvent.recipient": { type: "TEXT" },
    "atlasFeePayerEvent.rates": { type: "TEXT" },
    "atlasFeePayerEvent.funder": { type: "TEXT" },
    "atlasFeePayerEvent.tokenVault": { type: "TEXT" },
    "atlasFeePayerEvent.owningProfile": { type: "TEXT" },
    "atlasFeePayerEvent.tokenMint": { type: "TEXT" },
    "atlasFeePayerEvent.tokenOwner": { type: "TEXT" },
    "atlasFeePayerEvent.funderProfile": { type: "TEXT" },
    "atlasFeePayerEvent.funderKey": { type: "TEXT" },
    "atlasFeePayerEvent.funderVaultAuthority": { type: "TEXT" },
    "atlasFeePayerEvent.funderVault": { type: "TEXT" },
    "atlasFeePayerEvent.instructionSysvar": { type: "TEXT" },
    "atlasFeePayerEvent.vaultProgram": { type: "TEXT" },
    "atlasFeePayerEvent.funderTokenAccount": { type: "TEXT" },
    "atlasFeePayerEvent.owningKey": { type: "TEXT" },
    "atlasFeePayerEvent.tokenLimit": { type: "NUMERIC" },
    "atlasFeePayerEvent.conversionRate": { type: "NUMERIC" },
    "atlasFeePayerEvent.funderKeyIndex": { type: "NUMERIC" },
    "atlasFeePayerEvent.expectedConversionRateLower": { type: "NUMERIC" },
    "atlasFeePayerEvent.expectedConversionRateUpper": { type: "NUMERIC" },
    "atlasFeePayerEvent.signerCount": { type: "NUMERIC" },
    "atlasFeePayerEvent.keyIndex": { type: "NUMERIC" },
    "atlasFeePayerEvent.amount": { type: "NUMERIC" },
    "namedAccountsEvent.funder": { type: "TEXT" },
    "namedAccountsEvent.account": { type: "TEXT" },
    "namedAccountsEvent.name": { type: "TEXT" },
    "profileVaultEvent.profile": { type: "TEXT" },
    "profileVaultEvent.key": { type: "TEXT" },
    "profileVaultEvent.vault": { type: "TEXT" },
    "profileVaultEvent.vaultAuthority": { type: "TEXT" },
    "profileVaultEvent.tokensTo": { type: "TEXT" },
    "profileVaultEvent.fundsTo": { type: "TEXT" },
    "profileVaultEvent.funder": { type: "TEXT" },
    "profileVaultEvent.keyIndex": { type: "NUMERIC" },
    "profileVaultEvent.vaultSeed": { type: "TEXT" },
    "profileVaultEvent.amount": { type: "NUMERIC" },
    "sageEvent.outOfEscrowAmount": { type: "NUMERIC" },
    "sageEvent.tokenTransferAmount": { type: "NUMERIC" },
    "sageEvent.amountIntoEscrow": { type: "NUMERIC" },
    "sageEvent.burnAmount": { type: "NUMERIC" },
    "sageEvent.withdrawAmount": { type: "NUMERIC" },
    "sageEvent.depositAmount": { type: "NUMERIC" },
    "sageEvent.gameAndProfile_key": { type: "TEXT" },
    "sageEvent.gameAndProfile_profile": { type: "TEXT" },
    "sageEvent.gameAndProfile_gameId": { type: "TEXT" },
    "sageEvent.gameState": { type: "TEXT" },
    "sageEvent.funder": { type: "TEXT" },
    "sageEvent.sector1": { type: "TEXT" },
    "sageEvent.sector2": { type: "TEXT" },
    "sageEvent.ownerProfile": { type: "TEXT" },
    "sageEvent.ownerKey": { type: "TEXT" },
    "sageEvent.invalidator": { type: "TEXT" },
    "sageEvent.subProfile": { type: "TEXT" },
    "sageEvent.fleet": { type: "TEXT" },
    "sageEvent.gameAccounts_gameId": { type: "TEXT" },
    "sageEvent.gameAccounts_gameState": { type: "TEXT" },
    "sageEvent.sagePlayerProfile": { type: "TEXT" },
    "sageEvent.originTokenAccount": { type: "TEXT" },
    "sageEvent.ship": { type: "TEXT" },
    "sageEvent.shipEscrowTokenAccount": { type: "TEXT" },
    "sageEvent.starbaseAndStarbasePlayer_starbase": { type: "TEXT" },
    "sageEvent.starbaseAndStarbasePlayer_starbasePlayer": { type: "TEXT" },
    "sageEvent.gameAccountsAndProfile_gameAndProfileAndFaction_key": { type: "TEXT" },
    "sageEvent.gameAccountsAndProfile_gameAndProfileAndFaction_profile": { type: "TEXT" },
    "sageEvent.gameAccountsAndProfile_gameAndProfileAndFaction_profileFaction": { type: "TEXT" },
    "sageEvent.gameAccountsAndProfile_gameAndProfileAndFaction_gameId": { type: "TEXT" },
    "sageEvent.gameAccountsAndProfile_gameState": { type: "TEXT" },
    "sageEvent.gameAccountsFleetAndOwner_gameFleetAndOwner_fleetAndOwner_key": { type: "TEXT" },
    "sageEvent.gameAccountsFleetAndOwner_gameFleetAndOwner_fleetAndOwner_owningProfile": { type: "TEXT" },
    "sageEvent.gameAccountsFleetAndOwner_gameFleetAndOwner_fleetAndOwner_owningProfileFaction": { type: "TEXT" },
    "sageEvent.gameAccountsFleetAndOwner_gameFleetAndOwner_fleetAndOwner_fleet": { type: "TEXT" },
    "sageEvent.gameAccountsFleetAndOwner_gameFleetAndOwner_gameId": { type: "TEXT" },
    "sageEvent.gameAccountsFleetAndOwner_gameState": { type: "TEXT" },
    "sageEvent.fleetShips": { type: "TEXT" },
    "sageEvent.fundsTo": { type: "TEXT" },
    "sageEvent.craftingInstance": { type: "TEXT" },
    "sageEvent.craftingProcess": { type: "TEXT" },
    "sageEvent.craftingFacility": { type: "TEXT" },
    "sageEvent.craftingProgram": { type: "TEXT" },
    "sageEvent.subProfileInvalidator": { type: "TEXT" },
    "sageEvent.newSubProfile": { type: "TEXT" },
    "sageEvent.craftingRecipe": { type: "TEXT" },
    "sageEvent.cargoPodTo": { type: "TEXT" },
    "sageEvent.cargoType": { type: "TEXT" },
    "sageEvent.cargoStatsDefinition": { type: "TEXT" },
    "sageEvent.tokenFrom": { type: "TEXT" },
    "sageEvent.tokenTo": { type: "TEXT" },
    "sageEvent.tokenMint": { type: "TEXT" },
    "sageEvent.cargoProgram": { type: "TEXT" },
    "sageEvent.craftableItem": { type: "TEXT" },
    "sageEvent.key": { type: "TEXT" },
    "sageEvent.playerProfile": { type: "TEXT" },
    "sageEvent.disbandedFleet": { type: "TEXT" },
    "sageEvent.oldGameState": { type: "TEXT" },
    "sageEvent.newGameState": { type: "TEXT" },
    "sageEvent.cargoPod": { type: "TEXT" },
    "sageEvent.craftingDomain": { type: "TEXT" },
    "sageEvent.cargoHold": { type: "TEXT" },
    "sageEvent.fuelTank": { type: "TEXT" },
    "sageEvent.ammoBank": { type: "TEXT" },
    "sageEvent.cargoPodFrom": { type: "TEXT" },
    "sageEvent.mint": { type: "TEXT" },
    "sageEvent.mineItem": { type: "TEXT" },
    "sageEvent.resource": { type: "TEXT" },
    "sageEvent.location": { type: "TEXT" },
    "sageEvent.starbase": { type: "TEXT" },
    "sageEvent.surveyDataUnitTracker": { type: "TEXT" },
    "sageEvent.surveyDataUnitTrackerSigner": { type: "TEXT" },
    "sageEvent.gameId": { type: "TEXT" },
    "sageEvent.atlasTokenFrom": { type: "TEXT" },
    "sageEvent.atlasTokenTo": { type: "TEXT" },
    "sageEvent.signer": { type: "TEXT" },
    "sageEvent.profile": { type: "TEXT" },
    "sageEvent.planet": { type: "TEXT" },
    "sageEvent.fuelTokenFrom": { type: "TEXT" },
    "sageEvent.pointsCategory": { type: "TEXT" },
    "sageEvent.pointsModifierAccount": { type: "TEXT" },
    "sageEvent.pointsProgram": { type: "TEXT" },
    "sageEvent.game": { type: "TEXT" },
    "sageEvent.sector": { type: "TEXT" },
    "sageEvent.discoverer": { type: "TEXT" },
    "sageEvent.star": { type: "TEXT" },
    "sageEvent.gameStateAndProfile_gameAndProfile_key": { type: "TEXT" },
    "sageEvent.gameStateAndProfile_gameAndProfile_profile": { type: "TEXT" },
    "sageEvent.gameStateAndProfile_gameAndProfile_gameId": { type: "TEXT" },
    "sageEvent.gameStateAndProfile_gameState": { type: "TEXT" },
    "sageEvent.profileFaction": { type: "TEXT" },
    "sageEvent.starbasePlayer": { type: "TEXT" },
    "sageEvent.destinationTokenAccount": { type: "TEXT" },
    "sageEvent.sduTokenFrom": { type: "TEXT" },
    "sageEvent.sduTokenTo": { type: "TEXT" },
    "sageEvent.foodTokenFrom": { type: "TEXT" },
    "sageEvent.foodMint": { type: "TEXT" },
    "sageEvent.sduCargoType": { type: "TEXT" },
    "sageEvent.foodCargoType": { type: "TEXT" },
    "sageEvent.instructionsSysvar": { type: "TEXT" },
    "sageEvent.recentSlothashes": { type: "TEXT" },
    "sageEvent.nextShip": { type: "TEXT" },
    "sageEvent.oldShip": { type: "TEXT" },
    "sageEvent.next": { type: "TEXT" },
    "sageEvent.fromStarbase": { type: "TEXT" },
    "sageEvent.toStarbase": { type: "TEXT" },
    "sageEvent.fromSector": { type: "TEXT" },
    "sageEvent.toSector": { type: "TEXT" },
    "sageEvent.statsDefinition": { type: "TEXT" },
    "sageEvent.fuelMint": { type: "TEXT" },
    "sageEvent.feeTokenFrom": { type: "TEXT" },
    "sageEvent.feeTokenTo": { type: "TEXT" },
    "sageEvent.feeMint": { type: "TEXT" },
    "sageEvent.keyIndex": { type: "NUMERIC" },
    "sageEvent.ownerKeyIndex": { type: "NUMERIC" },
    "sageEvent.sector1Index": { type: "NUMERIC" },
    "sageEvent.sector2Index": { type: "NUMERIC" },
    "cargoEvent.authority": { type: "TEXT" },
    "cargoEvent.signerOriginAccount": { type: "TEXT" },
    "cargoEvent.statsDefinition": { type: "TEXT" },
    "cargoEvent.cargoPod": { type: "TEXT" },
    "cargoEvent.cargoType": { type: "TEXT" },
    "cargoEvent.originTokenAccount": { type: "TEXT" },
    "cargoEvent.cargoTokenAccount": { type: "TEXT" },
    "cargoEvent.funder": { type: "TEXT" },
    "cargoEvent.mint": { type: "TEXT" },
    "cargoEvent.tokenMint": { type: "TEXT" },
    "cargoEvent.key": { type: "TEXT" },
    "cargoEvent.profile": { type: "TEXT" },
    "cargoEvent.oldCargoType": { type: "TEXT" },
    "cargoEvent.mintAuthority": { type: "TEXT" },
    "cargoEvent.destinationTokenAccount": { type: "TEXT" },
    "cargoEvent.originPodAuthority": { type: "TEXT" },
    "cargoEvent.newPodAuthority": { type: "TEXT" },
    "cargoEvent.destinationPodAuthority": { type: "TEXT" },
    "cargoEvent.originCargoPod": { type: "TEXT" },
    "cargoEvent.destinationCargoPod": { type: "TEXT" },
    "cargoEvent.cargoAmount": { type: "NUMERIC" },
    "cargoEvent.mintAmount": { type: "NUMERIC" },
    "craftingEvent.burnAmount": { type: "NUMERIC" },
    "craftingEvent.key": { type: "TEXT" },
    "craftingEvent.profile": { type: "TEXT" },
    "craftingEvent.funder": { type: "TEXT" },
    "craftingEvent.recipe": { type: "TEXT" },
    "craftingEvent.domain": { type: "TEXT" },
    "craftingEvent.mint": { type: "TEXT" },
    "craftingEvent.recipeCategory": { type: "TEXT" },
    "craftingEvent.craftingFacility": { type: "TEXT" },
    "craftingEvent.craftableItem": { type: "TEXT" },
    "craftingEvent.location": { type: "TEXT" },
    "craftingEvent.authority": { type: "TEXT" },
    "craftingEvent.craftingProcess": { type: "TEXT" },
    "craftingEvent.tokenFrom": { type: "TEXT" },
    "craftingEvent.tokenTo": { type: "TEXT" },
    "craftingEvent.fundsTo": { type: "TEXT" },
    "craftingEvent.signer": { type: "TEXT" },
    "craftingEvent.newProfile": { type: "TEXT" },
    "craftingEvent.recipeCategoryOld": { type: "TEXT" },
    "craftingEvent.recipeCategoryNew": { type: "TEXT" },
    "craftingEvent.recipeDurationOverride": { type: "NUMERIC" },
    "raffleEvent.raffle": { type: "TEXT" },
    "raffleEvent.tickets": { type: "TEXT" },
    "raffleEvent.prizes": { type: "TEXT" },
    "raffleEvent.proceeds": { type: "TEXT" },
    "raffleEvent.proceedsMint": { type: "TEXT" },
    "raffleEvent.authority": { type: "TEXT" },
    "raffleEvent.payer": { type: "TEXT" },
    "raffleEvent.source": { type: "TEXT" },
    "raffleEvent.deposit": { type: "TEXT" },
    "raffleEvent.mint": { type: "TEXT" },
    "raffleEvent.metadata": { type: "TEXT" },
    "raffleEvent.edition": { type: "TEXT" },
    "raffleEvent.sourceRecord": { type: "TEXT" },
    "raffleEvent.depositRecord": { type: "TEXT" },
    "raffleEvent.authRules": { type: "TEXT" },
    "raffleEvent.sysvarInstructions": { type: "TEXT" },
    "raffleEvent.authRulesProgram": { type: "TEXT" },
    "raffleEvent.ataProgram": { type: "TEXT" },
    "raffleEvent.user": { type: "TEXT" },
    "raffleEvent.tokenAccount": { type: "TEXT" },
    "raffleEvent.slotHashes": { type: "TEXT" },
    "raffleEvent.payout": { type: "TEXT" },
    "raffleEvent.solPaid": { type: "NUMERIC" },
    "raffleEvent.solRecovered": { type: "NUMERIC" },
    "neoSwapEvent.swapDataAccount": { type: "TEXT" },
    "neoSwapEvent.signer": { type: "TEXT" },
    "neoSwapEvent.sysvarInstructions": { type: "TEXT" },
    "neoSwapEvent.splTokenProgram": { type: "TEXT" },
    "neoSwapEvent.splAtaProgram": { type: "TEXT" },
    "neoSwapEvent.itemFromDeposit": { type: "TEXT" },
    "neoSwapEvent.mint": { type: "TEXT" },
    "neoSwapEvent.nftMetadata": { type: "TEXT" },
    "neoSwapEvent.itemToDeposit": { type: "TEXT" },
    "neoSwapEvent.nftMasterEdition": { type: "TEXT" },
    "neoSwapEvent.ownerTokenRecord": { type: "TEXT" },
    "neoSwapEvent.destinationTokenRecord": { type: "TEXT" },
    "neoSwapEvent.authRulesProgram": { type: "TEXT" },
    "neoSwapEvent.authRules": { type: "TEXT" },
    "neoSwapEvent.user": { type: "TEXT" },
    "neoSwapEvent.leafDelegate": { type: "TEXT" },
    "neoSwapEvent.treeAuthority": { type: "TEXT" },
    "neoSwapEvent.merkleTree": { type: "TEXT" },
    "neoSwapEvent.signerAta": { type: "TEXT" },
    "neoSwapEvent.userAta": { type: "TEXT" },
    "neoSwapEvent.root": { type: "TEXT" },
    "neoSwapEvent.dataHash": { type: "TEXT" },
    "neoSwapEvent.creatorHash": { type: "TEXT" },
    "neoSwapEvent.nonce": { type: "NUMERIC" },
    "neoSwapEvent.index": { type: "NUMERIC" },
    "neoSwapEvent.amount": { type: "NUMERIC" },
};
class StarAtlasEventStream extends base_data_stream_class_1.BaseDataStream {
    id = "StarAtlasEvent";
    static fieldMap = fieldMap;
}
exports.StarAtlasEventStream = StarAtlasEventStream;
