// @shyft-to/js/dist/esm/api/mp-bidding-client.js
import { restApiCall } from '../utils';
export class MpBiddingClient {
    constructor(config) {
        this.config = config;
    }
    async active(input) {
        var _a;
        try {
            const params = {
                network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                marketplace_address: input.marketplaceAddress,
            };
            if (input === null || input === void 0 ? void 0 : input.buyerAddress) {
                params['buyer_address'] = input.buyerAddress;
            }
            if (input === null || input === void 0 ? void 0 : input.collectionAddress) {
                params['collection_address'] = input.collectionAddress;
            }
            if (input === null || input === void 0 ? void 0 : input.nftAddress) {
                params['nft_address'] = input.nftAddress;
            }
            if (input === null || input === void 0 ? void 0 : input.sortBy) {
                params['sort_by'] = input.sortBy;
            }
            if (input === null || input === void 0 ? void 0 : input.sortOrder) {
                params['sort_order'] = input.sortOrder;
            }
            if (input === null || input === void 0 ? void 0 : input.page) {
                params['page'] = input.page;
            }
            if (input === null || input === void 0 ? void 0 : input.size) {
                params['size'] = input.size;
            }
            const data = await restApiCall(this.config.apiKey, {
                method: 'get',
                url: 'marketplace/active_bids',
                params,
            });
            const bids = data.result;
            bids.data.forEach((x) => {
                x.created_at = new Date(x.created_at);
            });
            return bids;
        }
        catch (error) {
            throw error;
        }
    }
    async bid(input) {
        var _a;
        try {
            const reqBody = {
                network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                marketplace_address: input.marketplaceAddress,
                nft_address: input.nftAddress,
                price: input.price,
                buyer_wallet: input.buyerWallet,
            };
            if (input === null || input === void 0 ? void 0 : input.isGasLess) {
                reqBody['on_the_house'] = input.isGasLess;
            }
            if (input === null || input === void 0 ? void 0 : input.serviceCharge) {
                reqBody['service_charge'] = input.serviceCharge;
            }
            const data = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'marketplace/bid',
                data: reqBody,
            });
            const response = data.result;
            return response;
        }
        catch (error) {
            throw error;
        }
    }
    async cancelBid(input) {
        var _a;
        try {
            const reqBody = {
                network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                marketplace_address: input.marketplaceAddress,
                bid_state: input.bidState,
                buyer_wallet: input.buyerWallet,
            };
            if (input === null || input === void 0 ? void 0 : input.isGasLess) {
                reqBody['on_the_house'] = input.isGasLess;
            }
            const data = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'marketplace/cancel_bid',
                data: reqBody,
            });
            const encodedTransaction = data.result.encoded_transaction;
            return encodedTransaction;
        }
        catch (error) {
            throw error;
        }
    }
    async acceptBid(input) {
        var _a;
        try {
            const reqBody = {
                network: (_a = input === null || input === void 0 ? void 0 : input.network) !== null && _a !== void 0 ? _a : this.config.network,
                marketplace_address: input.marketplaceAddress,
                bid_state: input.bidState,
                seller_wallet: input.sellerWallet,
            };
            const data = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'marketplace/accept_bid',
                data: reqBody,
            });
            const response = data.result;
            return response;
        }
        catch (error) {
            throw error;
        }
    }
}
