// @shyft-to/js/dist/esm/api/semi-custodial-wallet-client.js
import { restApiCall } from '../utils';
export class SemiCustodialWalletClient {
    constructor(config) {
        this.config = config;
    }
    async create(input) {
        try {
            const reqBody = {
                password: input.password,
            };
            const data = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'semi_wallet/create',
                data: reqBody,
            });
            const walletAddress = data.result.wallet_address;
            return walletAddress;
        }
        catch (error) {
            throw error;
        }
    }
    async getKeypair(input) {
        try {
            const params = {
                password: input.password,
                wallet: input.walletAddress,
            };
            const data = await restApiCall(this.config.apiKey, {
                method: 'get',
                url: 'semi_wallet/get_keypair',
                params,
            });
            const wallet = data.result;
            return wallet;
        }
        catch (error) {
            throw error;
        }
    }
    async changePassword(input) {
        try {
            const reqBody = {
                current_password: input.currentPassword,
                new_password: input.newPassword,
                wallet: input.walletAddress,
            };
            const data = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'semi_wallet/change_password',
                data: reqBody,
            });
            const isPasswordChanged = data.success;
            return isPasswordChanged;
        }
        catch (error) {
            throw error;
        }
    }
}
