// @shyft-to/js/dist/esm/api/storage-client.js
import FormData from 'form-data';
import { restApiCall } from '../utils';
export class StorageClient {
    constructor(config) {
        this.config = config;
    }
    async uploadAsset(input) {
        try {
            let data = new FormData();
            data.append('file', input.file);
            const response = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'storage/upload',
                maxBodyLength: Infinity,
                data,
            });
            const uploadedAsset = response.result;
            return uploadedAsset;
        }
        catch (error) {
            throw error;
        }
    }
    async createMetadata(input) {
        try {
            const reqBody = {
                creator: input.creator,
                image: input.image,
                name: input.name,
                symbol: input.symbol,
                description: input.description,
                attributes: input.attributes,
                share: 100,
            };
            if (input === null || input === void 0 ? void 0 : input.external_url) {
                reqBody['external_url'] = input.external_url;
            }
            if (input === null || input === void 0 ? void 0 : input.sellerFeeBasisPoints) {
                reqBody['royalty'] = input.sellerFeeBasisPoints;
            }
            if (input === null || input === void 0 ? void 0 : input.files) {
                reqBody['files'] = input.files;
            }
            const data = await restApiCall(this.config.apiKey, {
                method: 'post',
                url: 'metadata/create',
                data: reqBody,
            });
            const uploadedMetadata = data.result;
            return uploadedMetadata;
        }
        catch (error) {
            throw error;
        }
    }
}
